README.md
# LibVXShare

An unofficial API wrapper for VirusShare.com

[![Build Status](https://travis-ci.org/chapinb/libvxshare.svg?branch=master)](https://travis-ci.org/chapinb/libvxshare) 
[![Coverage Status](https://coveralls.io/repos/github/chapinb/libvxshare/badge.svg?branch=master)](https://coveralls.io/github/chapinb/libvxshare?branch=master) 
[![MIT Licence](https://badges.frapsoft.com/os/mit/mit.svg?v=103)](https://opensource.org/licenses/mit-license.php)

## Installation

`pip install libvxshare`

## Documentation

* Module documentation: https://chapinb.com/libvxshare
* Official documentation: https://virusshare.com/apiv2_reference


ChangeLog.md
# LibVXShare Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2020-08-11

### Added

* Initial support for VirusShare API
* Implemented time-based rate limiting
* Unit test coverage of the API wrapper
* Documentation of the API


libvxshare/tests/test_apiv2.py
import datetime
import unittest
from unittest.mock import patch, Mock

from requests import HTTPError

from libvxshare.apiv2 import VirusShare


class TestAPIv2(unittest.TestCase):
    def setUp(self):
        self.sample_data = {}
        self.requests_get_path = 'libvxshare.apiv2.requests.get'

    def mock_json(self):
        return self.sample_data

    def test_file_exists(self):
        self.sample_data = {"response": 1}
        vxs = VirusShare(api_key='not_an_api_key')
        with patch(self.requests_get_path) as mock_get:
            mock_get.return_value.status_code = 200
            mock_get.return_value.json = self.mock_json
            actual = vxs.file_exists("15d5c88dd0f15eb4426e191b91f2cefe")
        expected = self.sample_data.copy()
        expected.update(dict(_detected=True, _exists=True))
        self.assertDictEqual(expected, actual)

    def test_file_report(self):
        self.sample_data = {
            "filetype": "Zip archive data, at least v2.0 to extract",
            "sha1": "778d0b014c58e8130804c56e9579a97eb77eee79",
            "ssdeep": "393216:Fap6O7g0jcBlDxonTFwEf0Hez/3+xpT4WWFobJ2p1ok:RKg0jcBlDxohzMHQupWFobY",
            "size": 16617145,
            "exif": {
                "FileSize": "16 MB",
                "FileType": "ZIP",
                "FileTypeExtension": "zip",
                "MIMEType": "application/zip",
                "ZipBitFlag": "0x0808",
                "ZipCRC": "0xc8373873",
                "ZipCompressedSize": 76009,
                "ZipCompression": "Deflated",
                "ZipFileName": "META-INF/MANIFEST.MF",
                "ZipModifyDate": "2018:11:05 16:54:10",
                "ZipRequiredVersion": 20,
                "ZipUncompressedSize": 226704
            },
            "sha512": "4c810dab39f0b8298ec383b1632aba77eb5c509e4ff28f7f7c09dd2dff69ada5be2a48afc80bec00064c6a215ce"
                      "52ec2f08a80b342dd7cd18fc6997a0555fd2a",
            "extension": "apk",
            "md5": "15d5c88dd0f15eb4426e191b91f2cefe",
            "added_timestamp": "2020-07-09T17:41:32+00:00",
            "trid": {
                "0": "Android Package (43.2%)",
                "1": "OpenOffice Extension (24.1%)",
                "2": "Java Archive (15.1%)",
                "3": "(.SH3D) Sweet Home 3D design (generic) (11.7%)",
                "4": "ZIP compressed archive (4.4%)"
            },
            "mimetype": "application/zip",
            "sha256": "036eef46db1f85a5c499d363e0f4d3b40051c9ef81ccd3f2173008c0c5dea4f4",
            "sha384": "e252519be8909e2b390fedd83f5b7d1317a19409b6ff329d54db329025636b6c8a9fedf391"
                      "512811268bfc389c40f509",
            "sha224": "d848a0d6f4971c09295b9275fbd4814933b297dbc7768d662035998a",
            "response": 1
        }
        vxs = VirusShare(api_key='not_an_api_key')
        with patch(self.requests_get_path) as mock_get:
            mock_get.return_value.status_code = 200
            mock_get.return_value.json = self.mock_json
            actual = vxs.file_report("15d5c88dd0f15eb4426e191b91f2cefe")
        expected = self.sample_data.copy()
        expected.update(dict(_detected=True, _exists=True))
        self.assertDictEqual(expected, actual)

    def test_file_source_convert(self):
        self.sample_data = {
            "sha256": "a1ac533baaf7de1dae53cf5b465aeca28a7f20bdfc79e5a0a39437dd728c231f",
            "size": 293376,
            "urllist": [
                {
                    "url": "http://totes-leg.it/bin/Data/Managed/Mono.Security.dll",
                    "timestamp": 1591933584
                },
                {
                    "url": "http://totes-leg.it/railmaker.apk:assets/bin/Data/Managed/Mono.Security.dll",
                    "timestamp": 1591976721
                }
            ]
        }
        vxs = VirusShare(api_key="not_an_api_key")
        with patch(self.requests_get_path) as mock_get:
            mock_get.return_value.status_code = 200
            mock_get.return_value.json = self.mock_json
            actual = vxs.file_source(
                hash_value="a1ac533baaf7de1dae53cf5b465aeca28a7f20bdfc79e5a0a39437dd728c231f",
                convert_timestamps=True
            )
        expected = self.sample_data.copy()
        expected.update(
            dict(urllist=[
                    {
                        "url": "http://totes-leg.it/bin/Data/Managed/Mono.Security.dll",
                        "timestamp": 1591933584,
                        "datetime": datetime.datetime(2020, 6, 12, 3, 46, 24)
                    },
                    {
                        "url": "http://totes-leg.it/railmaker.apk:assets/bin/Data/Managed/Mono.Security.dll",
                        "timestamp": 1591976721,
                        "datetime": datetime.datetime(2020, 6, 12, 15, 45, 21)
                    }
                 ]
            )
        )
        self.assertDictEqual(expected, actual)

    def test_file_source(self):
        self.sample_data = {
            "sha256": "a1ac533baaf7de1dae53cf5b465aeca28a7f20bdfc79e5a0a39437dd728c231f",
            "size": 293376,
            "urllist": [
                {
                    "url": "http://totes-leg.it/bin/Data/Managed/Mono.Security.dll",
                    "timestamp": 1591933584
                },
                {
                    "url": "http://totes-leg.it/railmaker.apk:assets/bin/Data/Managed/Mono.Security.dll",
                    "timestamp": 1591976721
                }
            ]
        }
        vxs = VirusShare(api_key="not_an_api_key")
        with patch(self.requests_get_path) as mock_get:
            mock_get.return_value.status_code = 200
            mock_get.return_value.json = self.mock_json
            actual = vxs.file_source(
                hash_value="a1ac533baaf7de1dae53cf5b465aeca28a7f20bdfc79e5a0a39437dd728c231f",
                convert_timestamps=False
            )
        expected = self.sample_data.copy()
        self.assertDictEqual(expected, actual)

    def test_source_md5(self):
        vxs = VirusShare()
        self.assertRaises(ValueError, vxs.file_source, "2c599b6155feb202941edf9bbf2d580d")

    def test_no_api_key(self):
        vxs = VirusShare()
        self.assertRaises(ValueError, vxs.file_exists, "2c599b6155feb202941edf9bbf2d580d")

    def test_raise_500(self):
        vxs = VirusShare(api_key="not_an_api_key")
        self.assertRaises(HTTPError, vxs.file_exists, "2c599b6155feb202941edf9bbf2d580d")

    def test_file_download(self):
        self.sample_data = b'PK\x03\x04\x14\x00\t\x00\x08'
        vxs = VirusShare(api_key='not_an_api_key')
        with patch(self.requests_get_path) as mock_get:
            mock_get.return_value.status_code = 200
            mock_get.return_value.content = self.sample_data
            actual = vxs.file_download("2c599b6155feb202941edf9bbf2d580d")
        expected = self.sample_data
        self.assertEqual(expected, actual)

    def test_rate_limit(self):
        vxs = VirusShare(api_key='not_an_api_key', requests_per_minute=30)
        with patch(self.requests_get_path) as mock_get:
            mock_get.side_effect = [
                Mock(status_code=204),
                Mock(status_code=200, content=self.sample_data)
            ]
            start = datetime.datetime.now()
            actual = vxs.file_download("2c599b6155feb202941edf9bbf2d580d")
        expected = self.sample_data
        self.assertGreaterEqual((datetime.datetime.now() - start).total_seconds(), 2)
        self.assertEqual(expected, actual)


if __name__ == '__main__':
    unittest.main()


searchindex.js
Search.setIndex({docnames:["APIv2","index"],envversion:{"sphinx.domains.c":1,"sphinx.domains.changeset":1,"sphinx.domains.citation":1,"sphinx.domains.cpp":1,"sphinx.domains.index":1,"sphinx.domains.javascript":1,"sphinx.domains.math":2,"sphinx.domains.python":1,"sphinx.domains.rst":1,"sphinx.domains.std":1,sphinx:56},filenames:["APIv2.rst","index.rst"],objects:{"":{libvxshare:[1,0,0,"-"]},"libvxshare.apiv2":{VirusShare:[0,1,1,""]},"libvxshare.apiv2.VirusShare":{file_download:[0,2,1,""],file_exists:[0,2,1,""],file_report:[0,2,1,""],file_source:[0,2,1,""]},libvxshare:{apiv2:[0,0,0,"-"]}},objnames:{"0":["py","module","Python module"],"1":["py","class","Python class"],"2":["py","method","Python method"]},objtypes:{"0":"py:module","1":"py:class","2":"py:method"},terms:{"byte":0,"case":0,"class":0,"return":0,"try":[0,1],That:1,The:0,Will:0,about:0,access:1,across:1,add:0,adding:1,allow:[0,1],alreadi:1,ani:1,api:1,api_kei:0,apiv2:0,apiv2_refer:[0,1],archiv:0,around:[0,1],associ:0,autom:1,becaus:1,befor:0,benign:0,call:0,can:1,claus:0,code:[0,1],com:[0,1],come:1,confirm:0,contain:0,content:[0,1],convers:1,convert:0,convert_timestamp:0,counter:1,data:[0,1],dataset:0,datetim:0,detect:0,dictionari:0,document:[0,1],doe:1,download:0,eas:1,endpoint:0,entri:0,error:0,exampl:0,except:[0,1],exist:[0,1],fals:0,field:0,file:0,file_download:0,file_exist:0,file_report:0,file_sourc:0,flexibl:0,found:0,framework:1,from:[0,1],fulli:0,gather:0,gener:1,github:1,greater:0,handl:[0,1],has:0,hash:0,hash_valu:0,helper:1,here:[0,1],highli:0,how:0,http:[0,1],implement:[0,1],includ:[0,1],index:[0,1],inform:[0,1],integr:1,interest:0,interfac:1,json:0,kei:0,leverag:[0,1],librari:[0,1],libvxshar:0,like:0,limit:[0,1],locat:0,mai:0,main:0,map:0,md5:0,metadata:0,method:0,minut:1,modul:[0,1],more:1,must:0,name:1,none:0,note:1,object:0,offici:1,one:0,onli:1,open:1,oper:0,other:1,page:[0,1],paramet:0,password:0,per:1,pleas:[0,1],plu:0,point:0,process:1,properli:0,protect:0,provid:1,python:1,queri:[0,1],quick:0,quickli:0,rais:[0,1],rate:[0,1],recommend:0,remain:1,report:0,request:[0,1],requests_per_minut:0,requir:1,respons:[0,1],said:1,same:1,sampl:0,script:1,search:[0,1],self:0,set:0,sha1:0,sha224:0,sha256:0,sha384:0,sha512:0,shown:0,simpl:1,soon:1,sourc:0,statu:0,stream:0,string:0,success:[0,1],support:[0,1],than:1,thi:[0,1],thread:1,ticket:1,timestamp:0,type:1,urllist:0,use:0,util:1,valu:0,version:0,virusshar:1,where:0,whether:0,which:0,within:[0,1],would:0,wrap:1,wrapper:1,you:0,your:1,zip:0},titles:["VirusShare API v2","LibVXShare"],titleterms:{api:0,indic:[0,1],libvxshar:1,overview:1,tabl:[0,1],virusshar:0}})

libvxshare/__init__.py
"""
Overview
=========

VirusShare.com API wrapper

LibVXShare includes a simple Python interface that wraps the already simple VirusShare API, namely adding support for
automated rate limiting and conversion of types in to Python data types. More helpers coming soon! Open a ticket
on GitHub with any requests.

Official API documentation here: https://virusshare.com/apiv2_reference

This library provides a generic interface for querying and accessing information from VirusShare's API. This can
allow for ease of integration in to existing Python frameworks, scripts, and other utilities. That said, it will
raise an exception for HTTP response codes other than 200 (success) and 204 (rate limiting). Please leverage
try/except around the library to handle exceptions as required by your implementation.

Notes
^^^^^

This implementation will not rate limit across threads or processes, only within the same thread. This is because the
API is limited on requests per minute and does not provide a counter of remaining requests.

"""
__name__ = "libvxshare"
__desc__ = "Unofficial VirusShare API Wrapper"
__version__ = "1.0.0"
__author__ = "Chapin Bryce"
__email__ = 'python@chapinb.com'
__license__ = "MIT"
__date__ = "20200810"


libvxshare/apiv2.py
"""
VirusShare API v2
=================

Support for version 2 of the VirusShare API.

This API is fully documented here: https://virusshare.com/apiv2_reference

An exception is raised in the case the status code is not 200 (Success) or 204 (Rate limited).

Please add try/except clauses around calls to this library to properly handle errors in requesting data.
This was not implemented within the library to allow greater flexibility in how you would like to handle errors.

"""
from time import sleep
import datetime

import requests


class VirusShare:
    """Class containing methods to support querying the VirusShare API including rate limiting operations."""
    def __init__(self, api_key=None, requests_per_minute=4):
        """Initialize class

        :param requests_per_minute: Integer representing the number of requests per minute allowed by your API
        """
        self.api_key = api_key
        self.uri = "https://virusshare.com/apiv2"
        self.requests_per_minute = requests_per_minute
        self.rate_limit_buffer = 0.01

        self._last_request = None
        self._rate_limit = float(60.0 / requests_per_minute) + self.rate_limit_buffer

    def _sleep(self):
        """Provide rate limiting based on API requests per minute.

        This executes immediately before each request to the API

        :return: None
        """
        if self._last_request is None:
            self._last_request = datetime.datetime.now()
            return

        elapsed_time = (datetime.datetime.now() - self._last_request).total_seconds()
        # If less than the rate limit has passed, sleep for the remainder
        if elapsed_time < self._rate_limit:
            sleep(self._rate_limit - elapsed_time)

    def _request(self, endpoint, hash_value):
        """Execute the API request and return the result.

        Incorporates the rate limiter.

        :param endpoint (str): A valid API endpoint. One of file, quick, download, source

        :return: Response from GET request.
        """
        if not self.api_key:
            raise ValueError("Required API key not provided")
        self._sleep()
        resp = requests.get("{}/{}".format(self.uri, endpoint), params={
            "apikey": self.api_key,
            "hash": hash_value
        })
        # If successful, return the results
        if resp.status_code == 200:
            return resp
        # If rate limited, sleep then try again
        if resp.status_code == 204:
            self._sleep()
            return self._request(endpoint, hash_value)

        # Otherwise raise the HTTP error
        resp.raise_for_status()

    @staticmethod
    def _add_metadata(response_data):
        """Add interpretation of the `response` value.

        0 - Indicates the file is not found
        1 - Indicates the file is found and detected
        2 - Indicated the file is found and benign

        :param response_data: Dictionary containing at least the 'response' key
        :return:
        """
        response_data['_exists'] = True if response_data.get('response', 0) in (1, 2) else False
        response_data['_detected'] = True if response_data.get('response', 0) == 1 else False
        return response_data

    def file_exists(self, hash_value):
        """Quickly confirm whether the file exists in the data set or not.

        This leverages the `/quick` endpoint, and will return a 0 if the file is not found,
        a 1 if the file is found and has a detection, and 2 if the file is found and benign.

        :param hash_value: A string hash value. May be one of md5, sha1, sha224, sha256, sha384, or sha512.
        :return: Dictionary containing response from VirusShare, plus metadata fields about whether the file exists
            or was detected.
        """
        response = self._request('quick', hash_value)
        resp_data = response.json()
        resp_data = self._add_metadata(resp_data)
        return resp_data

    def file_report(self, hash_value):
        """Gather a report about a file in the dataset.

        It is highly recommended to use self.file_exists to confirm the file is found
        within the dataset before requesting a report.

        This leverages the `/file` endpoint, which returns a JSON document with key-value mappings of
        the metadata associated with this entry. An example response is shown in the main documentation
        here: https://virusshare.com/apiv2_reference.

        :param hash_value: A string hash value. May be one of md5, sha1, sha224, sha256, sha384, or sha512.
        :return: Dictionary containing response from VirusShare, plus metadata fields about the request
        """
        response = self._request('file', hash_value)
        resp_data = response.json()
        resp_data = self._add_metadata(resp_data)
        return resp_data

    def file_source(self, hash_value, convert_timestamps=False):
        """Gather information about where the sample was sourced from.

        It is highly recommended to use self.file_exists to confirm the file is found
        within the dataset before requesting a report.

        This leverages the `/source` endpoint, which returns a JSON document with key-value mappings of
        the metadata associated with this entry. The main data point of interest is located in the
        `urllist` field. An example response is shown in the main documentation
        here: https://virusshare.com/apiv2_reference.

        :param hash_value: A string hash value. Must by SHA256
        :param convert_timestamps: Whether or not to convert timestamps to datetime objects. Will add `datetime` field.
        :return: Dictionary containing response from VirusShare, plus metadata fields about the request
        """
        if len(hash_value) != 64:
            raise ValueError("Please provide a SHA256 value")
        response = self._request('source', hash_value)
        resp_data = response.json()
        if convert_timestamps and isinstance(resp_data.get('urllist'), list) and len(resp_data.get('urllist')) > 0:
            new_url_list = []
            for entry in resp_data.get('urllist', []):
                new_url_list.append({
                    "url": entry.get("url"),
                    "timestamp": entry.get("timestamp"),
                    "datetime": datetime.datetime.utcfromtimestamp(entry.get("timestamp"))
                })
            resp_data['urllist'] = new_url_list
        return resp_data

    def file_download(self, hash_value):
        """Download a sample by hash value.

        It is highly recommended to use self.file_exists to confirm the file is found
        within the dataset before requesting a report.

        This leverages the `/download` endpoint, which returns a byte stream containing a password
        protected zip file archive.

        :param hash_value: A string hash value. May be one of md5, sha1, sha224, sha256, sha384, or sha512.
        :return: Byte content of the zip file containing the sample.
        """
        response = self._request('download', hash_value)
        return response.content
